### ISSUE TITLE
Pages wrapped with React.memo don’t work if concurrentFeatures is enabled

### ISSUE THREAD

**Issue by davecardwell** [vercel/next.js#31993]
### What version of Next.js are you using?

12.0.4 / 12.0.5-canary.12

### What version of Node.js are you using?

16.11.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

next dev

### Describe the Bug

If you enable `experimental.concurrentFeatures` and export a page component wrapped with `React.memo` you hit the following error:

> error - Error: Your page must export a `default` component

### Expected Behavior

No error, page uses React.memo.

### To Reproduce

See https://github.com/davecardwell/nextjs-memo-page-repro for reproduction, or follow the steps below:

```sh
yarn create next-app
# modify pages/index.js to wrap default export in React.memo()
yarn dev
# confirm that the page renders
yarn add react@beta react-dom@beta
# modify next.config.js to set experimental.concurrentFeatures to true
yarn dev
# confirm that the page does not render and you see the following error in the terminal:
# error - Error: Your page must export a `default` component
```

**Comment by davecardwell** [comment]
Thanks @huozhi! I had been experimenting with [eslint-plugin-react-memo](https://www.npmjs.com/package/eslint-plugin-react-memo) and came across the error. I don’t have a particular use case for memo-ing the page, but reported it as I thought it might be a symptom of an invalid check that might miss other actual issues.

**Comment by balazsorban44** [comment]
@davecardwell I am almost certain that that plugin is actively harmful... :eyes: Calling those optimization methods is not a guarantee in the first place, but calling them everywhere also doesn't come for free, and they should be used sparingly. Enforcing it on an entire codebase could backfire.

https://reactjs.org/docs/hooks-reference.html#usememo

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

https://reactjs.org/docs/react-api.html#reactmemo

> This method only exists as a [performance optimization](https://reactjs.org/docs/optimizing-performance.html). Do not rely on it to “prevent” a render, as this can lead to bugs.

**Comment by davecardwell** [comment]
@balazsorban44 Indeed. I found [the blog article](https://attardi.org/why-we-memo-all-the-things/) that inspired the plugin somewhat compelling, hence the experiment, but ultimately landed on the side of selectively using memo when noticing a performance issue as recommended by the React team.
